Fixes #37 - Buffers in multiple groups is now supported

Compro-Prasad opened the pull request:
Switching groups will now show buffers from that group only, even if a buffer is
present in multiple groups. Previously this would fail to work if a buffer is in
multiple groups.

Compro-Prasad commented on centaur-tabs.el:
```diff
@@ -1791,7 +1800,7 @@ Run as `centaur-tabs-quit-hook'."
   (interactive)
   (centaur-tabs-select-beg-tab t))
 
-(defun centaur-tabs-select-beg-tab (&optional backward _type)
```
I removed `_type` because it wasn't being used in the function.

Compro-Prasad commented:
Hello @ema2159 could you please test this?
I am currently using https://github.com/Compro-Prasad/emacs-config/blob/4579394a57f14da1efdb7250222f7ee8e15e1477/init.el#L862-L886 as a group function.

ema2159 commented:
Yes, give me just a little time, I've been busy so I've just been able to check minor issues!

ema2159 commented:
Sorry, the thing is I caught a throat infection so I've been in bed and I've only been able to answer stuff through my phone

Compro-Prasad commented:
No problem. I think I am too excited about taking `centaur-tabs` to the next level.

ema2159 commented:
> No problem. I think I am too excited about taking `centaur-tabs` to the next level.

I sincerely appreciate your help. You've helped me a lot with your issues/PRs.

Compro-Prasad commented:
Please take a look here. This adds a feature that was supposed to be there but it isn't. If we can have a list of groups representing a buffer, we should also be able to select any group and the group tabs display logic should be correct. Without this PR, many users wouldn't get this feature which is a very good addition. Consider it like, you can use mode groups and project groups at the same time. No need to change the group function everytime.

ema2159 commented:
Sorry, I forgot to try out this PR. I like the idea, but I encountered one issue, I move a lo with tab forward/backward movements (I have the cycle-scope variable set to nil). The thing is that, with this sometimes I'm unable to jump to another group with tab forward/backward movements, do you know why this is?

Compro-Prasad commented:
I actually don't use that feature. When cycling we should change the window parameter which sets the group for current window. This should probably fix the issue.

Compro-Prasad commented:
Still busy. Will find a time slot to work on this.

Compro-Prasad commented:
There are some bugs that I found out while reading the code. I want to mention them here and fix when I get time.
1. `centaur-tabs-selected-tab` function returns the first selected tab, not the current tab in current window. This might be a bug. Try opening same group in different windows having different selected tabs. Should probably be fixed using `current-buffer` or `window-parameter` (if `current-buffer` is unable to solve it).
2. `centaur-tabs-cycle` tries to do too much. I want to break it into three functions for the three conditions. Wrapping will be enabled by default. Configuration will be provided to disable wrapping.
  1. `centaur-tabs-cycle-tabs` to cycle only tabs.
  2. `centaur-tabs-cycle-groups` to cycle only groups. If one group, cycle tabs.
  3. `centaur-tabs-cycle-default` to cycle tabs first then groups if we hit the edge.
3. Remove `centaur-tabs-tab-next` function as it is too simple(doesn't have any knowledge about current context).

Compro-Prasad commented:
I have looked into this. And it seems that it is going to be a lot of work. I will have to update how groups are mapped to buffers.
I don't understand the need to update groups when switching tabs ( `centaur-tabs-current-tabset` accepts a `update` arg which seems to be useless as whenever this function is called, it is always passed a `t` arg which updates a lot of stuff by calling `centaur-tabs-buffer-update-groups` function. When switching tabs, this function gets called twice with `t` parameter passed even though it isn't required.). The sole problem is caused by the implementation of `centaur-tabs-buffer-update-groups`. I will have to isolate the situations, when to update and when not to update.

Compro-Prasad commented:
Pull and check if it works.

Compro-Prasad commented on centaur-tabs.el:
```diff
@@ -866,53 +914,18 @@ The scope of the cyclic navigation through tabs is specified by the
+  (let (;; If navigation through groups is requested, and there is
+	;; only one group, navigate through visible tabs.
+	(cycle
+	 (if (and (eq centaur-tabs-cycle-scope 'groups)
```
Why is this needed? IMO if an user prefers switching between groups then why should change the variable internally?

Compro-Prasad replied:
@ema2159 I asked this question to you. Is there anything you have to say about this?

ema2159 commented:
There's a lot of things from the code that are from Aquamacs's tabs implementation. We have to be really really careful making changes this big. They may cause unexpected bugs.

ema2159 commented:
Also, I need you to use Flycheck to verify there's no linting and compiling errors when submitting your PR's! You usually submit pretty nice and ordered code so let's try to keep it that way. I have incredibly high coding standards so let's try to eliminate all possible warnings and keep the nice little green check in the mode-line (if not, my OCD will be heavily triggered jaja).

Compro-Prasad commented:
That is why I asked you to test it properly before merging. After all this is my first package contribution that is this big. Besides, I have fixed what you asked for. Also, this fixes some edge cases that weren't handled before. There is a bug that I am unable to fix. This PR isn't a reason for the bug though. It wasn't there in the first place. When I fixed this multiple grouping issue, this new bug appeared. It wont affect users who don't use shared groups per buffer.
The bug: When you open two same buffers in different windows having different groups, the modified marker doesn't sync between them.

ema2159 commented:
I found a little bug. When doing `centaur-tabs-group-by-projectile-project` or `centaur-tabs-group-buffer groups`, the tabs should be updated and displayed immediately but for some reason they get the update after another centaur tabs command is inputed

ema2159 commented:
@Compro-Prasad don't worry,  you're doing a pretty good job, you've helped me a lot testing and fixing a lot of bugs with the package and I really appreciate your help. I'll try to test this thoroughly to ensure that the package keeps it's quality and I'll inform you about all the bugs I encounter!

Compro-Prasad commented:
The one bug remaining is currently not possible to fix and isn't a big issue. Its just modified marker not getting synced. I will try fixing that another time. For now, we need to find out other bugs.

ema2159 commented:
@Compro-Prasad cool. It seems that it's working fine now and all the linting issues are gone (almost, two minor issues remain). I'll try to test it thoroughly and when I feel it's alright I'll merge it. Thanks for all your great help and enthusiasm with the package!.

Compro-Prasad commented:
> By the way, just out of curiosity, where are you from?

India.

ema2159 commented:
> > By the way, just out of curiosity, where are you from?
> 
> India.

Awesome! It's been pretty nice to work with you

ema2159 commented:
Hi @Compro-Prasad! Are you still using this functionality? I'm currently thinking on reverting this given that as far as I know, it's not very used by other users and has caused some unexpected bugs like #62 and #58. Please let me know what you think.

Compro-Prasad commented:
I am using `tab-line` from Emacs 27. So, its okay to remove this feature.
